miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultFileAttachmentService.java

```diff
@@ -37,71 +41,108 @@
+    entityProviders.put("run", runService);
   }
 
-  public void setServiceRecordService(ServiceRecordService serviceRecordService) {
-    this.serviceRecordService = serviceRecordService;
+  @Autowired
+  public void setSampleService(SampleService sampleService) {
+    entityProviders.put("sample", sampleService);
   }
 
-  public void setAuthorizationManager(AuthorizationManager authorizationManager) {
-    this.authorizationManager = authorizationManager;
+  @Autowired
+  public void setServiceRecordService(ServiceRecordService serviceRecordService) {
+    entityProviders.put("servicerecord", serviceRecordService);
   }
 
   public void setFileStorageDirectory(String fileStorageDirectory) {
     this.fileStorageDirectory = fileStorageDirectory;
   }
 
   @Override
-  public Attachable get(String entityType, long entityId) {
-    Function<Long, Attachable> fetcher = entityFetchers.get(entityType);
-    if (fetcher == null) {
+  public Attachable get(String entityType, long entityId) throws IOException {
+    ProviderService<? extends Attachable> provider = entityProviders.get(entityType);
+    if (provider == null) {
       throw new IllegalArgumentException("Unknown entity type: " + entityType);
     }
-    return fetcher.apply(entityId);
+    return provider.get(entityId);
   }
 
   @Override
   public void add(Attachable object, MultipartFile file, AttachmentCategory category) throws IOException {
     Attachable managed = attachableStore.getManaged(object);
-    String saveFilename = Long.toString(new Date().getTime());
     String relativeDir = makeRelativeDir(object.getAttachmentsTarget(), object.getId());
+    File targetFile = storeFile(relativeDir, file);
+
+    try {
+      FileAttachment attachment = makeAttachment(file, relativeDir, targetFile, category);
+      managed.getAttachments().add(attachment);
+      attachableStore.save(managed);
+    } catch (Exception e) {
+      if (!targetFile.delete()) {
+        log.error("Failed to save attachment, but file was still saved: {}", targetFile.getAbsolutePath());
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public void addShared(Collection<Attachable> objects, MultipartFile file, AttachmentCategory category) throws IOException {
+    List<Attachable> managed = objects.stream().map(attachableStore::getManaged).collect(Collectors.toList());
+    if (managed.stream().map(Attachable::getAttachmentsTarget).distinct().count() > 1L) {
+      throw new IllegalArgumentException("Target objects are not all the same type");
+    }
+    String relativeDir = makeRelativeSharedDir(managed.get(0).getAttachmentsTarget());
+    File targetFile = storeFile(relativeDir, file);
+
+    try {
+      FileAttachment attachment = makeAttachment(file, relativeDir, targetFile, category);
+      for (Attachable item : managed) {
+        item.getAttachments().add(attachment);
+        attachableStore.save(item);
+      }
+    } catch (Exception e) {
+      if (!targetFile.delete()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Use "java.nio.Files#delete" here for better messages on error conditions. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4042)